sdk/cosmos/azure_data_cosmos/src/resource_context.rs

```diff
@@ -0,0 +1,233 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+use url::Url;
+
+use crate::utils::url_encode;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[allow(dead_code)] // For the variants. Can be removed when we have them all implemented.
+pub enum ResourceType {
+    Databases,
+    Containers,
+    Items,
+    StoredProcedures,
+    Users,
+    Permissions,
+    PartitionKeyRanges,
+    UserDefinedFunctions,
+    Triggers,
+    Offers,
+}
+
+impl ResourceType {
+    pub fn path_segment(self) -> &'static str {
+        match self {
+            ResourceType::Databases => "dbs",
+            ResourceType::Containers => "colls",
+            ResourceType::Items => "docs",
+            ResourceType::StoredProcedures => "sprocs",
+            ResourceType::Users => "users",
+            ResourceType::Permissions => "permissions",
+            ResourceType::PartitionKeyRanges => "pkranges",
+            ResourceType::UserDefinedFunctions => "udfs",
+            ResourceType::Triggers => "triggers",
+            ResourceType::Offers => "offers",
+        }
+    }
+}
+
+/// Represents a "resource link" defining a sub-resource in Azure Cosmos DB
+///
+/// This value is URL encoded, and can be [`Url::join`]ed to the endpoint root to produce the full absolute URL for a Cosmos DB resource.
+/// It's also intended for use by the signature algorithm used when authenticating with a primary key.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ResourceLink {
+    parent: Option<String>,
+    item_id: Option<String>,
+    resource_type: ResourceType,
+}
+
+impl ResourceLink {
+    pub fn root(resource_type: ResourceType) -> Self {
+        Self {
+            parent: None,
+            resource_type,
+            item_id: None,
+        }
+    }
+
+    pub fn feed(&self, resource_type: ResourceType) -> Self {
```

Comment:
If this is exported and customers are meant to construct these (or can), builder setters should use `with_feed`. It's also recommended in guidelines to take `mut self` in builder methods since references can cause issues in some cases e.g., you can't just "stop" in the middle until you call the final `build()`.

Reply:
These aren't quite builder APIs in the same was as usual. It's more of an immutable value that allows you to quickly create a clone "child" value. The idea is the `DatabaseClient` holds a `ResourceLink` for the database and then uses these APIs to create the child link to give to a `ContainerClient`.

The APIs take references because they are intended to leave the original `ResourceLink` alone and create a completely new `ResourceLink` that is "descended" from `self`. Also, there's no final `build` step, every returned `ResourceLink` is complete enough to be used.